lint/rule.go

```diff
@@ -14,10 +14,172 @@
 
 package lint
 
+import (
+	"fmt"
+	"strings"
+
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+)
+
 // Rule defines a lint rule that checks Google Protobuf APIs.
-type Rule interface {
-	// Info returns metadata about a rule.
-	Info() RuleInfo
-	// Lint performs the linting process.
-	Lint(Request) ([]Problem, error)
+type Rule struct {
```

Comment:
This is a very major change. We have 20 rules in google3 that this will break. Can you propose a new struct with a different name that conforms to the Rule interface? It should be relatively easy to do and prevent the need for a bunch of rework.

```go
type LukesRule struct {
	// Name contains the rule's name.
	Info() RuleInfo		Name RuleName
	// Lint performs the linting process.	
 	Lint(Request) ([]Problem, error)		// Description is a short description of the rule,
	// used in documentation and elsewhere.
	Description string

 	// URI is the address where the guideline is documented.
	// This should be displayed to give API designers more information about
	// "how to do this right".
	URI string

 	// LintFile is called for files as a whole.
	LintFile func(*desc.FileDescriptor) []Problem

 	// LintMessage is called for any messages in the file.
	// This includes nested messages, regardless of depth.
	LintMessage func(*desc.MessageDescriptor) []Problem

 	// LintField is called for any fields in the file.
	// This includes fields within nested messages, regardless of depth.
	LintField func(*desc.FieldDescriptor) []Problem

 	// LintService is called for each service in the file.
	LintService func(*desc.ServiceDescriptor) []Problem

 	// LintMethod is called for each method in the file.
	LintMethod func(*desc.MethodDescriptor) []Problem

 	// LintEnum is called for each enum in the file.
	LintEnum func(*desc.EnumDescriptor) []Problem
}

func (l LukesRule) Info() RuleInfo {
  return RuleInfo{
    Name: l.RuleName,
    Description: l.Description,
    URI: l.URI,
  }
}

func (l LukesRule) Lint(r Request) ([]Problem, error) {
  // logic for traversing callbacks.
}
```

otherwise syncing will be blocked until someone refactors all the rules in google3

Reply:
Syncing will be blocked anyway, considering the other changes.

I am willing to refactor the rules in google3 (we just need to stop making more of them). It was not difficult to refactor the ones here.

Reply:
@robcapo and I discussed this in email, and decided that I will take responsibility for migrating their existing rules (and that they will not create new ones until this can go in).